storm-core/src/jvm/backtype/storm/blobstore/FileBlobStoreImpl.java

```diff
@@ -0,0 +1,248 @@
+    }
+  }
+
+  /**
+   * @return all keys that are available for reading.
+   * @throws IOException on any error.
+   */ 
+  public Iterator<String> listKeys() throws IOException {
+    return new KeyInHashDirIterator();
+  }
+
+  /**
+   * Get an input stream for reading a part.
+   * @param key the key of the part to read.
+   * @return the where to read the data from.
+   * @throws IOException on any error
+   */
+  public LocalFsBlobStoreFile read(String key) throws IOException {
+    return new LocalFsBlobStoreFile(getKeyDir(key), BlobStoreFile.BLOBSTORE_DATA_FILE);
+  }
+
+  /**
+   * Get an object tied to writing the data.
+   * @param key the key of the part to write to.
+   * @return an object that can be used to both write to, but also commit/cancel the operation.
+   * @throws IOException on any error
+   */ 
+  public LocalFsBlobStoreFile write(String key, boolean create) throws IOException {
+    return new LocalFsBlobStoreFile(getKeyDir(key), true, create);
+  }
+
+  /**
+   * Check if the key exists in the blob store.
+   * @param key the key to check for
+   * @return true if it exists else false.
+   */
+  public boolean exists(String key) {
+    return getKeyDir(key).exists();
+  }
+  
+  /**
+   * Delete a key from the blob store
+   * @param key the key to delete
+   * @throws IOException on any error
+   */
+  public void deleteKey(String key) throws IOException {
+    File keyDir = getKeyDir(key);
+    LocalFsBlobStoreFile pf = new LocalFsBlobStoreFile(keyDir, BlobStoreFile.BLOBSTORE_DATA_FILE);
+    pf.delete();
+    delete(keyDir);
+  }
+
+  private File getKeyDir(String key) {
+    String hash = String.valueOf(Math.abs((long)key.hashCode()) % BUCKETS);
+    File ret = new File(new File(_fullPath, hash), key);
+    LOG.debug("{} Looking for {} in {}", new Object[]{_fullPath, key, hash});
+    return ret;
+  }
+
+  public void fullCleanup(long age) throws IOException {
+    long cleanUpIfBefore = System.currentTimeMillis() - age;
+    Iterator<String> keys = new KeyInHashDirIterator();
+    while (keys.hasNext()) {
+      String key = keys.next();
+      File keyDir = getKeyDir(key);
+      Iterator<LocalFsBlobStoreFile> i = listBlobStoreFiles(keyDir);
+      if (!i.hasNext()) {
+        //The dir is empty, so try to delete it, may fail, but that is OK
+        try {
+          keyDir.delete();
+        } catch (Exception e) {
+          LOG.warn("Could not delete "+keyDir+" will try again later");
+        }
+      }
+      while (i.hasNext()) {
+        LocalFsBlobStoreFile f = i.next();
+        if (f.isTmp()) {
+          if (f.getModTime() <= cleanUpIfBefore) {
+            f.delete();
+          }
+        }
+      }
+    }
+  }
+
+  protected Iterator<LocalFsBlobStoreFile> listBlobStoreFiles(File path) throws IOException {
+    ArrayList<LocalFsBlobStoreFile> ret = new ArrayList<LocalFsBlobStoreFile>();
+    File[] files = path.listFiles();
+    if (files != null) {
```

Comment:
File#listFiles does not return null.  It can return an empty array, but we handle that.